### PR TITLE
feat: remove redundant `EnableSideEffects` instructions

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

## Problem\*

Resolves 

## Summary\*

We currently pick up quite a lot of duplicated `EnableSideEffects` instructions as we perform optimizations on the SSA which tends to clutter up the output. This PR adds a small change to remember the currently active predicate condition and skips any instructions which will just will just reactivate the current predicate.

I've added a test case which shows the sort of SSA which we want to avoid.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs [hunk 1]
```diff
@@ -62,7 +63,8 @@ impl Context {
     ) {
         let instructions = function.dfg[block].take_instructions();
 
-        let mut last_side_effects_enabled_instruction = None;
+        let mut active_condition = function.dfg.make_constant(FieldElement::one(), Type::bool());
+        let mut last_side_effects_enabled_instruction: Option<InstructionId> = None;
```
Comment (TomAFrench) [inline compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs:67]:
```suggestion
        let mut last_side_effects_enabled_instruction = None;
```

File: compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs [hunk 2]
```diff
@@ -172,3 +181,88 @@ impl Context {
...
+        assert_eq!(instructions.len(), 9);
+
+        // Expected output:
+        //
+        // acir(inline) fn main f0 {
+        //     b0(v0: Field):
+        //       v3 = mul v0, Field 2
+        //       v4 = mul v0, Field 2
+        //       v5 = mul v0, Field 2
+        //       v6 = mul v0, Field 2
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs:253]: I was looking through the rest of SSA and it looks like we never added a pass to remove equivalent instructions (in an equivalent `EnableSideEffectsIf` context). I could see this improving some programs

**Review by jfecher — APPROVED** [review]
LGTM

File: compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs (hunk 2 above)
Comment (TomAFrench) [reply compiler/noirc_evaluator/src/ssa/opt/remove_enable_side_effects.rs:253]: This should be handled by the constant-folding pass tbh.

**Pull request merged**